controller_manager/src/controller_manager.cpp

```diff
@@ -137,6 +128,141 @@ void ControllerManager::getControllerNames(std::vector<std::string> &names)
   }
 }
 
+void ControllerManager::manageSwitch(const ros::Time &time)
+{
+  // switch hardware interfaces (if any)
+  if (!switch_started_)
+  {
+    robot_hw_->doSwitch(switch_start_list_, switch_stop_list_);
+    switch_started_ = true;
+  }
+
+  stopControllers(time);
+
+  // start controllers once the switch is fully complete
+  if (!start_asap_)
+  {
+    startControllers(time);
+  }
+  // start controllers as soon as their required joints are done switching
+  else
+  {
+    startControllersAsap(time);
+  }
+}
+
+void ControllerManager::stopControllers(const ros::Time &time)
+{
+  // stop controllers
+  for (auto &request : stop_request_)
+  {
+    if (request->isRunning())
+    {
+      if (!request->stopRequest(time))
+      {
+        ROS_FATAL("Failed to stop controller in realtime loop. This should never happen.");
+      }
+    }
+  }
+}
+
+void ControllerManager::startControllers(const ros::Time &time)
+{
+  // start controllers
+  if (robot_hw_->switchResult() == hardware_interface::RobotHW::DONE)
+  {
+    for (auto &request : start_request_)
+    {
+      if (!request->startRequest(time))
+      {
+        ROS_FATAL("Failed to start controller in realtime loop. This should never happen.");
+      }
+    }
+
+    please_switch_ = false;
+    switch_started_ = false;
+  }
+  // abort controllers in case of error or timeout (if set)
+  else if ((robot_hw_->switchResult() == hardware_interface::RobotHW::ERROR) ||
+           (timeout_ > 0.0 && (time - init_switch_).toSec() > timeout_))
+  {
+    for (auto &request : start_request_)
+    {
+      if (!request->abortRequest(time))
+      {
+        ROS_FATAL("Failed to abort controller in realtime loop. This should never happen.");
+      }
+    }
+
+    please_switch_ = false;
```

Comment:
I'm wondering if these variables could be refactored into a struct/class to allow for structured resetting (or other operations)

Reply:
I think that makes sense. I'll see to it.

Reply:
Done!